File: src/domain/userProfile.ts
import { DEFAULT_PAYDAY } from './period'
import type { Halalas } from './money'

/**
 * ملف المستخدم — قسم «الحساب» وأسئلة البداية (OVERRIDES §26، قرار المالك 2026-09-13).
 *
 * - **يوم الراتب لكل مستخدم** والفترات المالية بتتحسب منه. لو مش محدد ⇒ 28 (المالك).
 *   الأيام 29–31 بتتقيد بآخر يوم في الشهر في `period.ts` (`clampPaydayToMonth`) مش هنا.
 * - **الاسم والمرتب والنوع و«بيعول حد» اختياريين:** `null` = ما اتجاوبش، ومش «لأ».
 *   (قاعدة 10: لا صفر مؤكد مكان المجهول — مرتب مش متسجل ≠ مرتب صفر.)
 * - المرتب عدد صحيح بالهللة (قاعدة 1).
 * - **معلومات بتظهر تصنيفات (OVERRIDES §28.1):** عنده سيارة، بيعول مين، ساكن بإيجار، عنده عمالة
 *   منزلية، عنده شغل خاص. `null` = ما اتجاوبش ⇒ التصنيف المشروط مخفي.
 */

export type Gender = 'male' | 'female'

/** «بتعول مين؟» — الزوج أو الزوجة أو الأولاد بيظهروا تصنيف «الأسرة والأطفال». */
export type DependentKind = 'spouse' | 'children' | 'parents'
export const DEPENDENT_KINDS: readonly DependentKind[] = ['spouse', 'children', 'parents']

export interface UserProfile {
  displayName: string | null
  salaryMinor: Halalas | null
  payday: number
  gender: Gender | null
  /** هل بيعول حد. */
  supportsDependents: boolean | null
  /** بيعول مين بالظبط؛ `null` = ما اتجاوبش. بيتسأل بس لو «بيعول حد» = أيوه. */
  dependentKinds: DependentKind[] | null
  hasCar: boolean | null
  renter: boolean | null
  domesticWorker: boolean | null
  business: boolean | null
  /**
   * وقت ما خلّص أسئلة البداية؛ null = لسه ما خلصهاش ⇒ الأسئلة بتظهر له
   * (أي حساب، جديد أو قديم — OVERRIDES §26).
   */
  onboardedAt: string | null
}

export const MAX_NAME_LENGTH = 60

export function emptyProfile(): UserProfile {
  return {
    displayName: null, salaryMinor: null, payday: DEFAULT_PAYDAY, gender: null, supportsDependents: null,
    dependentKinds: null, hasCar: null, renter: null, domesticWorker: null, business: null, onboardedAt: null,
  }
}

export type ProfileField =
  | 'displayName' | 'salaryMinor' | 'payday' | 'gender' | 'supportsDependents'
  | 'dependentKinds' | 'hasCar' | 'renter' | 'domesticWorker' | 'business'

export type ProfileCheck =
  | { ok: true; profile: UserProfile }
  | { ok: false; field: ProfileField; message: string }

const YES_NO_FIELDS = ['supportsDependents', 'hasCar', 'renter', 'domesticWorker', 'business'] as const

const isDependentKind = (value: unknown): value is DependentKind =>
  typeof value === 'string' && (DEPENDENT_KINDS as readonly string[]).includes(value)

/**
 * قراءة ملف متخزن بأمان: حقل ناقص أو بنوع غلط بيبقى «ما اتجاوبش» (أو 28 ليوم الراتب)
 * بدل ما يوقع فتح التطبيق. مش بيكتب حاجة — التصحيح بيحصل أول حفظ من المستخدم.
 */
export function parseStoredProfile(raw: unknown): UserProfile {
  const data = raw && typeof raw === 'object' ? (raw as Record<string, unknown>) : {}
  const profile = emptyProfile()
  if (typeof data.displayName === 'string' && data.displayName.trim() && data.displayName.trim().length <= MAX_NAME_LENGTH) {
    profile.displayName = data.displayName.trim()
  }
  if (Number.isSafeInteger(data.salaryMinor) && (data.salaryMinor as number) >= 0) profile.salaryMinor = data.salaryMinor as number
  if (Number.isInteger(data.payday) && (data.payday as number) >= 1 && (data.payday as number) <= 31) profile.payday = data.payday as number
  if (data.gender === 'male' || data.gender === 'female') profile.gender = data.gender
  for (const field of YES_NO_FIELDS) if (typeof data[field] === 'boolean') profile[field] = data[field] as boolean
  if (Array.isArray(data.dependentKinds) && data.dependentKinds.every(isDependentKind)) {
    profile.dependentKinds = DEPENDENT_KINDS.filter((kind) => (data.dependentKinds as unknown[]).includes(kind))
  }
  if (typeof data.onboardedAt === 'string' && data.onboardedAt) profile.onboardedAt = data.onboardedAt
  return profile
}

/** بيتأكد من قيم الملف قبل الحفظ، وبيرجّع الخطأ بجانب حقله (spec/04). */
export function checkProfile(input: UserProfile): ProfileCheck {
  const name = input.displayName === null ? null : input.displayName.trim()
  if (name !== null && name.length > MAX_NAME_LENGTH) {
    return { ok: false, field: 'displayName', message: `الاسم أطول من ${MAX_NAME_LENGTH} حرف` }
  }
  if (input.salaryMinor !== null && (!Number.isSafeInteger(input.salaryMinor) || input.salaryMinor < 0)) {
    return { ok: false, field: 'salaryMinor', message: 'المرتب لازم يكون مبلغ صحيح مش سالب' }
  }
  if (!Number.isInteger(input.payday) || input.payday < 1 || input.payday > 31) {
    return { ok: false, field: 'payday', message: 'يوم الراتب لازم يكون من 1 لـ 31' }
  }
  if (input.gender !== null && input.gender !== 'male' && input.gender !== 'female') {
    return { ok: false, field: 'gender', message: 'اختار ذكر أو أنثى، أو سيبها فاضية' }
  }
  for (const field of YES_NO_FIELDS) {
    if (input[field] !== null && typeof input[field] !== 'boolean') {
      return { ok: false, field, message: 'اختار أيوه أو لأ، أو سيبها فاضية' }
    }
  }
  if (input.dependentKinds !== null && (!Array.isArray(input.dependentKinds) || !input.dependentKinds.every(isDependentKind))) {
    return { ok: false, field: 'dependentKinds', message: 'اختار الزوج أو الزوجة أو الأولاد أو الأهل، أو سيبها فاضية' }
  }
  // «لأ» على بيعول حد ⇒ مفيش حد بيعوله؛ الترتيب ثابت ومن غير تكرار
  const dependentKinds = input.supportsDependents === false ? null
    : input.dependentKinds === null ? null
      : DEPENDENT_KINDS.filter((kind) => input.dependentKinds!.includes(kind))
  return { ok: true, profile: { ...input, displayName: name === '' ? null : name, dependentKinds } }
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/UserProfile.kt
package app.masroufy.core

/**
 * ملف المستخدم — نقل `src/domain/userProfile.ts` (OVERRIDES §26 و§28.1).
 * يوم الراتب لكل مستخدم (من غيره 28). الاختياري `null` = ما اتجاوبش، ومش «لأ» (قاعدة 10).
 */
data class UserProfile(
    val displayName: String?,
    val salaryMinor: Halalas?,
    val payday: Int,
    /** "male" أو "female". */
    val gender: String?,
    val supportsDependents: Boolean?,
    /** "spouse" / "children" / "parents" بترتيب ثابت. */
    val dependentKinds: List<String>?,
    val hasCar: Boolean?,
    val renter: Boolean?,
    val domesticWorker: Boolean?,
    val business: Boolean?,
    /** null = لسه ما خلّصش أسئلة البداية. */
    val onboardedAt: String?,
)

val DEPENDENT_KINDS = listOf("spouse", "children", "parents")
const val MAX_NAME_LENGTH = 60

fun emptyProfile() = UserProfile(null, null, DEFAULT_PAYDAY, null, null, null, null, null, null, null, null)

private fun safeInteger(value: Any?): Long? = when (value) {
    is Int -> value.toLong()
    is Long -> value.takeIf { it in -MAX_SAFE_HALALAS..MAX_SAFE_HALALAS }
    is Double -> value.takeIf { it % 1.0 == 0.0 && kotlin.math.abs(it) <= MAX_SAFE_HALALAS.toDouble() }?.toLong()
    else -> null
}

/** قراية ملف متخزن بأمان: الحقل الناقص أو الغلط بيبقى «ما اتجاوبش» (أو 28 ليوم الراتب) — مش بيكتب حاجة. */
fun parseStoredProfile(raw: Map<String, Any?>?): UserProfile {
    val data = raw ?: emptyMap()
    val name = (data["displayName"] as? String)?.let(JsText::trim)
    val salary = safeInteger(data["salaryMinor"])
    val payday = safeInteger(data["payday"])
    val kinds = data["dependentKinds"] as? List<*>
    fun bool(field: String) = data[field] as? Boolean
    return UserProfile(
        displayName = name?.takeIf { it.isNotEmpty() && it.length <= MAX_NAME_LENGTH },
        salaryMinor = salary?.takeIf { it >= 0 },
        payday = payday?.takeIf { it in 1..31 }?.toInt() ?: DEFAULT_PAYDAY,
        gender = (data["gender"] as? String)?.takeIf { it == "male" || it == "female" },
        supportsDependents = bool("supportsDependents"),
        dependentKinds = kinds?.takeIf { list -> list.all { it is String && it in DEPENDENT_KINDS } }?.let { list -> DEPENDENT_KINDS.filter { it in list } },
        hasCar = bool("hasCar"), renter = bool("renter"), domesticWorker = bool("domesticWorker"), business = bool("business"),
        onboardedAt = (data["onboardedAt"] as? String)?.takeIf { it.isNotEmpty() },
    )
}

sealed interface ProfileCheck {
    data class Ok(val profile: UserProfile) : ProfileCheck
    data class Invalid(val field: String, val message: String) : ProfileCheck
}

/** تأكيد قيم الملف قبل الحفظ — الخطأ جنب حقله (spec/04). */
fun checkProfile(input: UserProfile): ProfileCheck {
    val name = input.displayName?.let(JsText::trim)
    if (name != null && name.length > MAX_NAME_LENGTH) return ProfileCheck.Invalid("displayName", "الاسم أطول من $MAX_NAME_LENGTH حرف")
    if (input.salaryMinor != null && (input.salaryMinor > MAX_SAFE_HALALAS || input.salaryMinor < 0)) {
        return ProfileCheck.Invalid("salaryMinor", "المرتب لازم يكون مبلغ صحيح مش سالب")
    }
    if (input.payday < 1 || input.payday > 31) return ProfileCheck.Invalid("payday", "يوم الراتب لازم يكون من 1 لـ 31")
    if (input.gender != null && input.gender != "male" && input.gender != "female") {
        return ProfileCheck.Invalid("gender", "اختار ذكر أو أنثى، أو سيبها فاضية")
    }
    if (input.dependentKinds != null && !input.dependentKinds.all { it in DEPENDENT_KINDS }) {
        return ProfileCheck.Invalid("dependentKinds", "اختار الزوج أو الزوجة أو الأولاد أو الأهل، أو سيبها فاضية")
    }
    // «لأ» على بيعول حد ⇒ مفيش حد بيعوله؛ الترتيب ثابت ومن غير تكرار
    val kinds = if (input.supportsDependents == false || input.dependentKinds == null) null else DEPENDENT_KINDS.filter { it in input.dependentKinds }
    return ProfileCheck.Ok(input.copy(displayName = name?.ifEmpty { null }, dependentKinds = kinds))
}
